### ISSUE TITLE
skin_js resources are loading before js resources when params are used

### ISSUE THREAD

**Issue by ProxiBlue** [AOEpeople/Aoe_JsCssTstamp#15]
Found a perplexing issue, which may be due to later magento versions?

When activated, I found that skin_js resources are loaded first, then js resources. This of course causes major issues, as many required core files are not yet loaded, namely prototype (to just mention one)

What was produced:

```
<script type="text/javascript" src="http://sante2017.enjo.test/js/plugincompany/contactforms/lib/head.load.min.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/plugincompany/contactforms/pccforms.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/rwd/default/js/lib/modernizr.custom.min.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/rwd/default/js/lib/selectivizr.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/rwd/default/js/lib/matchMedia.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/rwd/default/js/lib/matchMedia.addListener.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/rwd/default/js/lib/enquire.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/rwd/default/js/app.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/rwd/default/js/lib/jquery.cycle2.min.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/rwd/default/js/lib/jquery.cycle2.swipe.min.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/rwd/default/js/slideshow.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/rwd/default/js/lib/imagesloaded.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/rwd/default/js/minicart.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/sante/default/js/santeApp.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/sante/default/js/jquery.bxslider.min.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/sante/default/js/bxslider_data_wrapper.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/sante/default/js/jquery.lazyloadxt.min.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/sante/default/js/jquery.lazyloadxt.bg.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/sante/default/js/underscore-min.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/sante/default/js/isotope.pkgd.min.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/sante/default/js/isotope-packery.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/sante/default/js/jquery.mb.YTPlayer.min.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/sante/default/js/html5videoplayer-min.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/sante/default/js/category-button.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/sante/default/js/prototype-jquery-conflict-fix.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/sante/default/js/jquery.matchHeight.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/sante/default/js/jquery.magnific-popup.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/base/default/nwdthemes/revslider/public/assets/js/jquery.themepunch.tools.min.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/skin/frontend/base/default/nwdthemes/revslider/public/assets/js/jquery.themepunch.revolution.min.js"></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/prototype/prototype.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/varien/js.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/varien/form.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/lib/jquery/noconflict.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/lib/ccard.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/prototype/validation.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/scriptaculous/builder.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/scriptaculous/effects.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/scriptaculous/dragdrop.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/scriptaculous/controls.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/scriptaculous/slider.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/mage/translate.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/mage/cookies.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/magestore/sociallogin/lightbox.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/magestore/sociallogin/sociallogin.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/advinstagram/instafeed.min.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/algoliasearch/internals/frontend/Function.prototype.bind.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/algoliasearch/internals/frontend/algoliaBundle.min.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/algoliasearch/internals/frontend/common.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/algoliasearch/instantsearch.js" 1></script>
<script type="text/javascript" src="http://sante2017.enjo.test/js/algoliasearch/autocomplete.js" 1></script>
```

I tracked this down to https://github.com/AOEpeople/Aoe_JsCssTstamp/blob/master/app/code/community/Aoe/JsCssTstamp/Block/Head.php#L340 simply ordering wrong.

Produced :

![selection_186](https://user-images.githubusercontent.com/4994260/37947727-290ef692-31bf-11e8-9905-808783db3e64.png)

Notice that the js resources has array key 1

Next the skin_js resources sort, and they get key 0.

![selection_187](https://user-images.githubusercontent.com/4994260/37947846-bea9071a-31bf-11e8-80ed-b68a64e17011.png)

The end result, having been merged later, is that js resources are loaded last, breaking site.

**Comment by sprankhub** [comment]
Weird. The ordering is okay for me with 1.9.3.8... I currently do not have the time for it, but this should probably be checked in a clean environment.

**Comment by ProxiBlue** [comment]
I am starting to suspect it may be a module that exists in the ecosystem, which causes the oddity.
My current bet is: plugincompany/contactforms

When I have a moment, I will test that thought.

**Comment by ProxiBlue** [comment]
It is indeed plugincompany/contactforms whichis causing the issue. With it removed, all seems fine.

Looking at the layout directives, I see they added a ```sort="last"``` into the addJs actions as params.

```
<reference name="head">
            <action method="addJs">
                <name>plugincompany/contactforms/lib/head.load.min.js</name>
                <params>sort="last"</params>
            </action>
            <action method="addJs">
                <name>plugincompany/contactforms/pccforms.js</name>
                <params>sort="last"</params>
            </action>
        </reference>
```

So that param is what is causing the sort odder, as it creates the main js to fall under key id 1 when sorted in the same noted routine in this ticket.

I had never seen such a param be used. I wonder if it is valid?

**Comment by sprankhub** [comment]
I guess the idea behind this is explained [here](https://stackoverflow.com/a/13089204/719023). It is a trick  to be at the end of the JS includes. If this construct mixes up the sort order, this has to be fixed in the `Aoe_JsCssTstamp` module IMHO.

**Comment by ProxiBlue** [comment]
I pinged the plugin company support to come have a look at this, but that last comment you gave is helpful.

I negated my PR, as I think I broke ability to have multiple array elements in $items. I am having a new look on how to solve differently.

**Comment by ProxiBlue** [comment]
Oh - I see - 

>  Magento will put items with "params" set after any items with no "params" set.

**Comment by ProxiBlue** [comment]
Ok, new PR.

This will allow multiple 'left over' array(s) of items to be reordered in the routine, after the default set was pulled.

Not much test cases to work on. With just the one set of ```sort='last'``` set of js includes, the ordering worked as expected, and they are even placed last. (but will not be last in the grander scope of things, since skin_js will follow those, but they will be last in the /js/ list.

I added another group, as from what was exampled in that link shared.

```
<default>
        <reference name="head">
            <action method="addJs">
                <name>plugincompany/contactforms/lib/head.load.min.js</name>
                <params>sort="last"</params>
            </action>
            <action method="addJs">
                <name>plugincompany/contactforms/pccforms.js</name>
                <params>sort="last"</params>
            </action>
            <action method="addJs">
                <name>plugincompany/contactforms/lib/111head.load.min.js</name>
                <params>sort="best-last"</params>
            </action>
            <action method="addJs">
                <name>plugincompany/contactforms/111pccforms.js</name>
                <params>sort="best-last"</params>
            </action>
        </reference>
    </default>
```
then also correctly resulted in an expected ordered list:

![selection_192](https://user-images.githubusercontent.com/4994260/37955964-a15105d8-31dc-11e8-8f90-b76565e60a7f.png)

I don't know if this will deal with all potentials, but sofar, works for me.

**Comment by milansimek** [comment]
Hi guys,

I indeed put the sort=last param at the JS includes in the layout xml to make sure the files are included last.

The reason for this is that many DIY web shop owners and some crappy themes either include jQuery somewhere halfway in the JS loading order, or they include it multiple times on a page. This generally happens when they include jQuery using an extension that has a name which is below ours when sorted alphabetically (which Magento does).

Since our extension checks whether jQuery is loaded, and loads it dynamically if needed, it causes issues if jQuery is included below our script(s).

If you're including jQuery on the top of the page using your page.xml or local.xml file, you should be able to just remove the sort=last attribute from our layout xml file.

Anyway, it would of course be good to fix this in your module also, because this practice is used in multiple extensions. (although I have to admit it's a bit hacky)

**Comment by milansimek** [comment]
By the way, on our own website we use a 'data-group=groupname' parameter to group different bundles of JS. That way the merged JS files don't get too big since Google Lighthouse doesn't like big clusters of JS much :)

**Comment by sprankhub** [comment]
Fixed with #17.

**Comment by sprankhub** [comment]
Thanks @ProxiBlue
